File: LeetCode/0318.maximum-product-of-word-lengths/maximum-product-of-word-lengths.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int size = words.size();
        int val[size + 5] = {0};
        int res = 0;
        for (int i = 0; i < size; i++){
            int length = words[i].size();
            for (int j = 0; j < length; j++){
                val[i] |= (1 << (words[i][j] - 'a'));
            }

            for (int j = 0; j < i; j++){
                if (!(val[i] & val[j])){
                    res = max(res, (int)(words[i].size() * words[j].size()));
                }
            }
        }

        return res;
    }
};

File: LeetCode/0318.maximum-product-of-word-lengths/maximum-product-of-word-lengths.java
class Solution {
    public int maxProduct(String[] words) {
        int size = words.length;
        int[] val = new int[size + 5];
        int res = 0;
        for (int i = 0; i < size; i++){
            char[] cstr = words[i].toCharArray();
            int length = cstr.length;
            for (int j = 0; j < length; j++){
                val[i] |= (1 << (cstr[j] - 'a'));
            }

            for (int j = 0; j < i; j++){
                if ((val[j] & val[i]) == 0){
                    res = Math.max(res, words[i].length() * words[j].length());
                }
            }
        }

        return res;
    }
}